pkg/reconciler/stats_reporter.go

```diff
@@ -0,0 +1,172 @@
+		stats.UnitDimensionless)
+	DeploymentReadyLatencyStat = stats.Int64(
+		DeploymentReadyLatencyName,
+		"Latency of successful knative serving deployment reconcile in milliseconds",
+		stats.UnitMilliseconds)
+	// Create the tag keys that will be used to add tags to our measurements.
+	// Tag keys must conform to the restrictions described in
+	// go.opencensus.io/tag/validate.go. Currently those restrictions are:
+	// - length between 1 and 255 inclusive
+	// - characters are printable US-ASCII
+	reconcilerTagKey = tag.MustNewKey("reconciler")
+	keyTagKey        = tag.MustNewKey("key")
+)
+
+func init() {
+	// Create views to see our measurements. This can return an error if
+	// a previously-registered view has the same name with a different value.
+	// View name defaults to the measure name if unspecified.
+	if err := view.Register(
+		&view.View{
+			Description: InstallSuccessCountStat.Description(),
+			Measure:     InstallSuccessCountStat,
+			Aggregation: view.Count(),
+			TagKeys:     []tag.Key{reconcilerTagKey, keyTagKey},
+		},
+		&view.View{
+			Description: InstallSuccessLatencyStat.Description(),
+			Measure:     InstallSuccessLatencyStat,
+			Aggregation: view.Distribution(10, 100, 1000, 10000, 30000, 60000), // Bucket boundaries are 10ms, 100ms, 1s, 10s, 30s and 60s.
+			TagKeys:     []tag.Key{reconcilerTagKey, keyTagKey},
+		},
+		&view.View{
+			Description: DeploymentReadyCountStat.Description(),
+			Measure:     DeploymentReadyCountStat,
+			Aggregation: view.Count(),
+			TagKeys:     []tag.Key{reconcilerTagKey, keyTagKey},
+		},
+		&view.View{
+			Description: DeploymentReadyLatencyStat.Description(),
+			Measure:     DeploymentReadyLatencyStat,
+			Aggregation: view.Distribution(10, 100, 1000, 10000, 30000, 60000), // Bucket boundaries are 10ms, 100ms, 1s, 10s, 30s and 60s.
+			TagKeys:     []tag.Key{reconcilerTagKey, keyTagKey},
+		},
+	); err != nil {
+		panic(err)
+	}
+}
+
+// StatsReporter defines the interface for sending metrics
+type StatsReporter interface {
+	// ReportKnativeServingReady reports the count and latency metrics for a reconcile operation
+	ReportInstallSuccess(resourceNamespace, resourceName string, duration time.Duration) error
+	ReportDeploymentReady(resourceNamespace, resourceName string, duration time.Duration) error
+}
+
+// Reporter holds cached metric objects to report metrics
+type reporter struct {
+	reconciler string
+	ctx        context.Context
+}
+
+// srKey is used to associate StatsReporters with contexts.
+type srKey struct{}
+
+// WithStatsReporter attaches the given StatsReporter to the provided context
+// in the returned context.
+func WithStatsReporter(ctx context.Context, sr StatsReporter) context.Context {
+	return context.WithValue(ctx, srKey{}, sr)
+}
+
+// GetStatsReporter attempts to look up the StatsReporter on a given context.
+// It may return null if none is found.
+func GetStatsReporter(ctx context.Context) StatsReporter {
```

Comment:
Is it necessary to associate StatsReporters with contexts? I believe contexts are intended for cross-process usage (https://golang.org/pkg/context/), but StatsReporters will only be used within the one reconciler controller.